src/client/session/test.rs

```diff
@@ -0,0 +1,518 @@
+use std::{future::Future, time::Duration};
+
+use bson::{doc, Bson};
+use futures::stream::StreamExt;
+
+use crate::{
+    error::Result,
+    options::{Acknowledgment, FindOptions, InsertOneOptions, ReadPreference, WriteConcern},
+    test::{EventClient, TestClient, CLIENT_OPTIONS, LOCK},
+    Client,
+    Collection,
+    RUNTIME,
+};
+
+/// Macro defining a closure that returns a future populated by an operation on the
+/// provided client identifier.
+macro_rules! client_op {
+    ($client:ident, $body:expr) => {
+        |$client| async move {
+            $body.await.unwrap();
+        };
+    };
+}
+
+/// Macro defining a closure that returns a future populated by an operation on the
+/// provided database identifier.
+macro_rules! db_op {
+    ($test_name:expr, $db:ident, $body:expr) => {
+        |client| async move {
+            let $db = client.database($test_name);
+            $body.await.unwrap();
+        }
+    };
+}
+
+/// Macro defining a closure that returns a future populated by an operation on the
+/// provided collection identifier.
+macro_rules! collection_op {
+    ($test_name:expr, $coll:ident, $body:expr) => {
+        |client| async move {
+            let $coll = client.database($test_name).collection($test_name);
+            $body.await.unwrap();
+        }
+    };
+}
+
+/// Macro that runs the provided function with each operation that uses a session.
+macro_rules! for_each_op {
```

Comment:
I'm little confused about `for_each_op` both because it only seems to be used once and because it doesn't seem to have anything that requires it to be a macro rather than just a function. I'm guessing that it was abstracted out in preparation for when we implement prose tests 4 and 5 as part of explicit sessions?

Reply:
>I'm guessing that it was abstracted out in preparation for when we implement prose tests 4 and 5 as part of explicit sessions?

Yep

>it doesn't seem to have anything that requires it to be a macro rather than just a function.

I was having a hard time defining a function that could get the various layers of `Fn` generic parameters to work right. It was easier to just do it by macro.

Reply:
Fair enough. It took me a bit of time to fully get what was going on here, but I think most of that is due to there still not being a way to have async closures, which would have obviated the need for the `collection_op!`, etc.